File: 3148-maximum-difference-score-in-a-grid/3148-maximum-difference-score-in-a-grid.cpp
class Solution {
public:

    int n,m;
    vector<vector<int>>dp;
    int ans = INT_MIN;

    int solution(int i,int j,vector<vector<int>>&grid){
        if(i>=n || j>=m) return INT_MIN;

        if(dp[i][j]!=-1) return dp[i][j];

        int best = grid[i][j];

        if(j+1 < m){
            int right = solution(i,j+1,grid);
            ans = max(ans, right - grid[i][j]);
            best = max(best, right);
        }

        if(i+1 < n){
            int down = solution(i+1,j,grid);
            ans = max(ans, down - grid[i][j]);    
            best = max(best, down);
        }

        return dp[i][j] = best;
    }

    int maxScore(vector<vector<int>>& grid) {
        n=grid.size();
        m=grid[0].size();
        dp.assign(n,vector<int>(m,-1));

        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                solution(i,j,grid);
            }
        }
        return ans;
    }
};

File: 3148-maximum-difference-score-in-a-grid/3148-maximum-difference-score-in-a-grid.java
class Solution {

    int n,m;
    int dp[][];
    int ans=Integer.MIN_VALUE;

    int solution(int i,int j,List<List<Integer>> grid){
        if(i>=n || j>=m) return Integer.MIN_VALUE;

        if(dp[i][j]!=-1) return dp[i][j];
        int best=grid.get(i).get(j);

        if(i+1<n){
            int bottom=solution(i+1,j,grid);
            int value=grid.get(i).get(j);
            ans=Math.max(ans,bottom-value);
            best=Math.max(best,bottom);
        }

        if(j+1<m){
            int right=solution(i,j+1,grid);
            int value=grid.get(i).get(j);
            ans=Math.max(ans,right-value);
            best=Math.max(best,right);
        }

        return dp[i][j]=best;
    }

    public int maxScore(List<List<Integer>> grid) {
        n=grid.size();
        m=grid.get(0).size();
        dp=new int[n][m];

        for(int i=0;i<n;i++){
            Arrays.fill(dp[i],-1);
        }

        for(int i=0;i<n;i++){
            Arrays.fill(dp[i],-1);
            for(int j=0;j<m;j++){
                solution(i,j,grid);
            }
        }
        return ans;
    }
}